src/clr-angular/data/datagrid/providers/column-reorder.service.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable, ViewContainerRef, ViewRef } from '@angular/core';
+import { Observable, Subject, BehaviorSubject } from 'rxjs';
+import { ColumnsService } from './columns.service';
+import { DatagridColumnChanges } from '../enums/column-changes.enum';
+import { ClrDragEvent } from '../../../utils/drag-and-drop/drag-event';
+import { first } from 'rxjs/operators';
+
+let nbColumnsGroup = 0;
+
+export interface ReorderRequest {
+  sourceOrder: number;
+  targetOrder: number;
+}
+
+export type ReorderAnimRequest = ReorderRequest & ClrDragEvent<ViewRef>;
+
+@Injectable()
+export class ColumnReorderService {
+  // the common group id that will be shared across
+  // Datagrids all reorder draggable and droppables
+  private _columnsGroupId: string;
+
+  constructor(private columnsService: ColumnsService) {
+    this._columnsGroupId = 'dg-column-group-' + nbColumnsGroup++;
+  }
+
+  orders: number[];
+
+  private _ordersChange: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  private _reorderRequested: Subject<ReorderRequest> = new Subject<ReorderRequest>();
+  private _reorderAnimRequested: Subject<ReorderAnimRequest> = new Subject<ReorderAnimRequest>();
+
+  containerRef: ViewContainerRef;
+
+  get columnsGroupId(): string {
+    return this._columnsGroupId;
+  }
+
+  get ordersChange(): Observable<boolean> {
+    return this._ordersChange.asObservable();
+  }
+
+  get reorderRequested(): Observable<ReorderRequest> {
+    return this._reorderRequested.asObservable();
+  }
+
+  get reorderAnimRequested(): Observable<ReorderAnimRequest> {
+    return this._reorderAnimRequested.asObservable();
+  }
+
+  // The following method is called by column ClrDatagridColumn when one column is dropped on another.
+  reorderViews(sourceView: ViewRef, targetView: ViewRef, dropEvent: ClrDragEvent<ViewRef>): void {
+    const sourceOrder = this.containerRef.indexOf(sourceView);
```

Comment:
Question: what do you think about changing the variable names to something more descriptive? e.g - dragColumn and dropColumn? I know under the hood its manipulating views but reading it as sourceOrder/targetOrder its less clear to me whats happening here.

Reply:
I have done following changes: 

sourceView -> dragColumnView
sourceOrder -> dragColumnOrder

targetView -> dropColumnView
targetOrder -> dropColumnOrder